### DISCUSSION TITLE
Filter by Location Radius

### DISCUSSION THREAD

**Discussion by Marcello1099** [Budibase/budibase#11527] [Help] [answered]
Hello,

I was wondering if there is a way of filtering by location radius. For instance, select city X, and radius of 10km.

I have the data in coordinates, and could easily map some cities with their central coordinates. However, I cannot seem to find a way to filter by location radius to include all coordinates that fall within it.

Thanks in advance!

**Comment by andz-bb** [comment, marked as answer]
hey @Marcello1099, for this you will need to use some maths to calculate the distance between the selected city and your other location coordinates. I've made a quick example app to demonstrate this - [you can try the finished app for yourself here](https://andz.budibase.app/app/proximity-improved#/locations) and I'll leave an app export below for you to see how it's set up in detail.

<img width="1103" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/471e5855-c5d7-435e-b800-e456cf1dc36a">

Here's the component structure for my screen:

<img width="287" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/7b56c32c-2c03-464e-8bb7-19984c1be8e6">

I have a Data Provider that gets all of my locations and then a second _Filtered Locations_ provider which uses the first provider as its data source.

I then have set up an Options Picker which is populated with the central coordinates of a few cities. I am using a custom options source for this, set up like so:
<img width="930" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/389ae76e-00dd-45bd-9bfb-b1b3e3c35215">


Now that I have my picker set up, I want to get a list of locations near to my selected city. To do this, I'm going to add an Update State action to the On Change actions of my picker. Here I am using JavaScript to return an array of _ids for location rows that are within a 10km radius of my selected city. The `calculateDistance` function here uses [the Haversine formula](https://en.wikipedia.org/wiki/Haversine_formula) to calculate the distance between the coords taking into account the curvature of the earth. I've found this to be pretty accurate in my testing, although there are other ways to calculate this that may be less computationally heavy.

```js
function calculateDistance(coords1, coords2) {
    // the haversine formula
    const R = 6371; // earth's radius in km

    const dLat = (coords2[0] - coords1[0]) * Math.PI / 180;
    const dLon = (coords2[1] - coords1[1]) * Math.PI / 180;

    const a = 
        Math.sin(dLat / 2) * Math.sin(dLat / 2) +
        Math.cos(coords1[0] * Math.PI / 180) * Math.cos(coords2[0] * Math.PI / 180) *
        Math.sin(dLon / 2) * Math.sin(dLon / 2);
    
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
    const distance = R * c;

    return distance; // returns distance in km
}

const locations = $("Locations Provider.Rows") //  my list of locations unfiltered

// city coordinates from my picker, formatted to [55.000,-7.000] for the function
const targetCoords = $("Field Value").split(',').map(Number); 

const nearbyLocations = locations.filter(x => 
    // return locations less than 10km from the selected city
    calculateDistance(targetCoords, [x.latitude, x.longitude]) <= 10
);

// only return a list of the relevant row _ids, rather than the whole row objects
return nearbyLocations.map(x => x._id)
```

So now every time I update the selected option, it will calculate the nearby location _ids and set that array to state. I can use that state variable to filter my second data provider which is then being used for my map and table. 

<img width="1173" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/bacc7ffe-d766-4f36-ac88-b336f00c7cde">

```js
return $("State.nearbyLocations") || $("Locations Provider.Rows").map(x => x._id)
```

I think I've covered everything here! I realise this relies fairly heavily on custom JavaScript - if you have any questions about any part of it I'm happy to answer them. I've included the app export below if you want to play around with it yourself. Hope this helps 😊

[proximity improved-export-1692198759887.tar.gz](https://github.com/Budibase/budibase/files/12360824/proximity.improved-export-1692198759887.tar.gz)

**Reply by Marcello1099** [reply]
This is amazing, thank you so much!
